File: 368. Largest Divisible Subset.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        n = len(nums)
        if n <= 1:
            return nums
        nums.sort()
        ret = [1] * n
        last = [-1] * n
        for i in range(n):
            for j in range(i+1, n):
                if ret[i] < ret[j] or nums[j] % nums[i] != 0:
                    continue
                ret[j] = ret[i] + 1
                last[j] = i
        
        i = ret.index(max(ret))
        r = []
        while i >= 0:
            r.append(nums[i])
            i = last[i]
        #r.sort()
        return r

File: 368. Largest Divisible Subset.go
func largestDivisibleSubset(nums []int) []int {
	n := len(nums)
	sort.Ints(nums)

	maxi, maxl := 0, 0
	dp := make([]int, n)
	pre := make([]int, n)

	for i := 0; i < n; i++ {
		if dp[i] == 0 {
			dp[i] = 1
			pre[i] = -1
		}
		for j := i + 1; j < n; j++ {
			if nums[j]%nums[i] == 0 && dp[j] < dp[i]+1 {
				dp[j] = dp[i] + 1
				pre[j] = i
			}
		}
		if dp[i] > maxl {
			maxl = dp[i]
			maxi = i
		}
	}

	res := make([]int, 0, maxl)
	for i := maxi; i >= 0; i = pre[i] {
		res = append(res, nums[i])
	}

	return res
}
